crates/mysten-network/src/server.rs

```diff
@@ -205,22 +211,132 @@ impl<M: MetricsCallbackProvider> ServerBuilder<M> {
     }
 }
 
-async fn tcp_listener_and_update_multiaddr<T: ToSocketAddrs>(
+async fn listen_and_update_multiaddr<T: ToSocketAddrs>(
     address: &Multiaddr,
     socket_addr: T,
-) -> Result<(Multiaddr, TcpListenerStream)> {
-    let (local_addr, incoming) = tcp_listener(socket_addr).await?;
+    tls_config: Option<ServerConfig>,
+) -> Result<(
+    Multiaddr,
+    impl Stream<Item = std::io::Result<TcpOrTlsStream>>,
+)> {
+    let listener = TcpListener::bind(socket_addr).await?;
+    let local_addr = listener.local_addr()?;
     let local_addr = update_tcp_port_in_multiaddr(address, local_addr.port());
-    Ok((local_addr, incoming))
+
+    let tls_acceptor = tls_config.map(|tls_config| TlsAcceptor::from(Arc::new(tls_config)));
+    let incoming = TcpOrTlsListener::new(listener, tls_acceptor);
+    let stream = async_stream::stream! {
+        loop {
+            yield incoming.accept().await;
+        }
+    };
+
+    Ok((local_addr, stream))
 }
 
-async fn tcp_listener<T: ToSocketAddrs>(address: T) -> Result<(SocketAddr, TcpListenerStream)> {
-    let listener = TcpListener::bind(address).await?;
-    let local_addr = listener.local_addr()?;
-    let incoming = TcpListenerStream::new(listener);
-    Ok((local_addr, incoming))
+pub struct TcpOrTlsListener {
+    listener: TcpListener,
+    tls_acceptor: Option<TlsAcceptor>,
 }
 
+impl TcpOrTlsListener {
+    fn new(listener: TcpListener, tls_acceptor: Option<TlsAcceptor>) -> Self {
+        Self {
+            listener,
+            tls_acceptor,
+        }
+    }
+
+    async fn accept(&self) -> std::io::Result<TcpOrTlsStream> {
+        let (stream, addr) = self.listener.accept().await?;
+        if self.tls_acceptor.is_none() {
+            return Ok(TcpOrTlsStream::Tcp(stream, addr));
+        }
+
+        // Determine whether new connection is TLS.
+        let mut buf = [0; 1];
+        stream.peek(&mut buf).await?;
+        if buf[0] == 0x16 {
```

Comment:
Chatgpt suggests to enhance the check to `if buf[0] == 0x16 && (buf[1] == 0x03 && (buf[2] >= 0x01 && buf[2] <= 0x04)) ...`

Reply:
And add a DEBUG log here?

Reply:
Why would that more-complicated check be better? It seems all we need to do here is switch on TLS or not-TLS, and we can then rely on the TlsAcceptor to do the proper, full TLS exchange.

log sounds good